source/extensions/quic_listeners/quiche/envoy_quic_client_stream.cc

```diff
@@ -0,0 +1,235 @@
+
+void EnvoyQuicClientStream::encodeTrailers(const Http::HeaderMap& trailers) {
+  ASSERT(!local_end_stream_);
+  local_end_stream_ = true;
+  ENVOY_STREAM_LOG(debug, "encodeTrailers: {}.", *this, trailers);
+  WriteTrailers(envoyHeadersToSpdyHeaderBlock(trailers), nullptr);
+}
+
+void EnvoyQuicClientStream::encodeMetadata(const Http::MetadataMapVector& /*metadata_map_vector*/) {
+  // Metadata Frame is not supported in QUIC.
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+void EnvoyQuicClientStream::resetStream(Http::StreamResetReason reason) {
+  // Higher layers expect calling resetStream() to immediately raise reset callbacks.
+  runResetCallbacks(reason);
+  Reset(envoyResetReasonToQuicRstError(reason));
+}
+
+void EnvoyQuicClientStream::switchStreamBlockState(bool should_block) {
+  ASSERT(FinishedReadingHeaders(),
+         "Upper stream buffer limit is reached before response body is delivered.");
+  if (should_block) {
+    sequencer()->SetBlockedUntilFlush();
+  } else {
+    ASSERT(read_disable_counter_ == 0, "readDisable called in between.");
+    sequencer()->SetUnblocked();
+  }
+}
+
+void EnvoyQuicClientStream::OnInitialHeadersComplete(bool fin, size_t frame_len,
+                                                     const quic::QuicHeaderList& header_list) {
+  quic::QuicSpdyStream::OnInitialHeadersComplete(fin, frame_len, header_list);
+  if (rst_sent()) {
+    return;
+  }
+  ASSERT(decoder() != nullptr);
+  ASSERT(headers_decompressed());
+  decoder()->decodeHeaders(quicHeadersToEnvoyHeaders(header_list), /*end_stream=*/fin);
+  if (fin) {
+    end_stream_decoded_ = true;
+  }
+  ConsumeHeaderList();
+}
+
+void EnvoyQuicClientStream::OnBodyAvailable() {
+  ASSERT(FinishedReadingHeaders());
+  ASSERT(read_disable_counter_ == 0);
+  ASSERT(!in_decode_data_callstack_);
+  in_decode_data_callstack_ = true;
+
+  Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+  // TODO(danzh): check Envoy per stream buffer limit.
+  // Currently read out all the data.
+  while (HasBytesToRead()) {
+    struct iovec iov;
+    int num_regions = GetReadableRegions(&iov, 1);
+    ASSERT(num_regions > 0);
+    size_t bytes_read = iov.iov_len;
+    Buffer::RawSlice slice;
+    buffer->reserve(bytes_read, &slice, 1);
+    ASSERT(slice.len_ >= bytes_read);
+    slice.len_ = bytes_read;
+    memcpy(slice.mem_, iov.iov_base, iov.iov_len);
+    buffer->commit(&slice, 1);
+    MarkConsumed(bytes_read);
+  }
+
+  // True if no trailer and FIN read.
+  bool finished_reading = IsDoneReading();
+  bool empty_payload_with_fin = buffer->length() == 0 && fin_received();
+  // If this call is triggered by an empty frame with FIN which is not from peer
+  // but synthesized by stream itself upon receiving HEADERS with FIN or
+  // TRAILERS, do not deliver end of stream here. Because either decodeHeaders
+  // already delivered it or decodeTrailers will be called.
+  bool skip_decoding = empty_payload_with_fin && (end_stream_decoded_ || !finished_reading);
```

Comment:
we have a test for this?   I think overall more unit tests for this file would be good